pkg/action/executor/nondisruptive_helper.go

```diff
@@ -0,0 +1,174 @@
+		return fmt.Errorf("failed operations for non-disruptive action")
+	}
+
+	// There should be exactly two pods
+	if len(pods) != newReplicas {
+		glog.Errorf("Found %d (not 2) pods by %s %s/%s", h.contKind, h.namespace, h.contName)
+		return fmt.Errorf("failed operations for non-disruptive action")
+	}
+
+	for _, p1 := range pods {
+		if p1.Name != h.targetPod {
+			h.podToClean = p1.Name
+			break
+		}
+	}
+
+	glog.V(4).Infof("The pod to clean up: %s", h.podToClean)
+
+	glog.V(2).Infof("Finished non-disruptive operations on %s %s/%s", h.contKind, h.namespace, h.contName)
+
+	return nil
+}
+
+// Cleans up the pod created in the operation for non-disruptive action support.
+// The pod, if exists, will be deleted and the replicas of the associated controller
+// will be set to one.
+func (h *NonDisruptiveHelper) CleanUp() {
+	if h.podToClean == "" { // No operations for non-disruptive actions
+		glog.V(4).Infof("No need to clean up for non-disruptive operations on %s %s/%s", h.contKind, h.namespace, h.contName)
+		return
+	}
+
+	glog.V(2).Infof("Started clean up for non-disruptive operations on %s %s/%s", h.contKind, h.namespace, h.contName)
+
+	// Wait for at least two pods ready before deleting the podToClean
+	glog.V(2).Infof("Wait for the target pod to be ready...")
+	err := util.RetryDuring(defaultCheckUpdateReplicaRetry, defaultCheckUpdateReplicaTimeout, defaultCheckUpdateReplicaSleep, func() error {
+		ctrlReplicas, ctrlReadyReplicas, err := autil.GetReplicaStatus(h.client, h.namespace, h.contName, h.contKind)
+		if err != nil {
+			return err
+		}
+
+		glog.V(2).Infof("replicas: %d, readyReplicas: %d", ctrlReplicas, ctrlReadyReplicas)
+
+		if ctrlReadyReplicas > 1 {
+			return nil
+		}
+
+		return fmt.Errorf("the operation for the target pod has not yet finished")
+	})
+
+	if err != nil {
+		glog.Errorf("Error updating replica to 1 for %s %s/%s: %s", h.contKind, h.namespace, h.contName, err)
+	}
+
+	// Delete podToClean
+	glog.V(4).Infof("Deleting pod: %s", h.podToClean)
+	err = autil.DeletePod(h.client, h.namespace, h.podToClean)
+
+	if err != nil {
+		glog.Errorf("Error while deleting pod %s: %s", h.podToClean, err)
+	}
+
+	// Set replica = 1
+	glog.V(4).Infof("Updating replicas to 1")
+	if err := h.updateReplicaWithRetry(1); err != nil {
+		glog.Errorf("Error updating replica to 1 for %s %s/%s: %s", h.contKind, h.namespace, h.contName, err)
+		return
+	}
+
+	glog.V(2).Infof("Finished clean up for non-disruptive operations on %s %s/%s", h.contKind, h.namespace, h.contName)
+}
+
+func (h *NonDisruptiveHelper) updateReplicaWithRetry(newReplicas int) error {
+	// With retrying several times
+	glog.V(2).Infof("updateReplicaWithRetry: newReplicas = %d", newReplicas)
+	err := util.RetryDuring(defaultUpdateReplicaRetry, defaultUpdateReplicaTimeout, defaultUpdateReplicaSleep, func() error {
+		return autil.UpdateReplicas(h.client, h.namespace, h.contName, h.contKind, newReplicas)
```

Comment:
In the cleanUp, before actually updateReplicas, it will be kind of safer if we can check that the current replicaNum setting  is still 2.  
As our action may take several minutes, user may modify the replicaNum during this period. So it will be safer to do a check first.

Reply:
fixed